### ISSUE TITLE
Finetuning keeps the base checkpoint's stale input_features: make_policy only refreshes them when empty (root cause of #2418; official smolvla_libero config contradicts its own normalizer)

### ISSUE THREAD

**Issue by easyrider11** [huggingface/lerobot#4517]
### Bug

When finetuning with `--policy.path=...`, `make_policy` refreshes `output_features` from the dataset **unconditionally**, but `input_features` **only when empty** ([`factory.py` L333-L335 @ `bf31dd7`](https://github.com/huggingface/lerobot/blob/bf31dd794ffb43a944ef4977da2f9350a3ddc960/src/lerobot/policies/factory.py#L333-L335)):

```python
cfg.output_features = {key: ft for key, ft in features.items() if ft.type is FeatureType.ACTION}
if not cfg.input_features:
    cfg.input_features = {key: ft for key, ft in features.items() if key not in cfg.output_features}
```

A config loaded from a pretrained checkpoint always has non-empty `input_features`, so a finetune onto a different embodiment **keeps the base model's feature spec forever**, while the normalization stats come from the new dataset. The asymmetry with `output_features` suggests this is accidental.

### Evidence: an official checkpoint contradicts itself

[`lerobot/smolvla_libero`](https://huggingface.co/lerobot/smolvla_libero) (finetuned from `smolvla_base` on `lerobot/libero`, per its own `train_config.json`):

| Artifact | `observation.state` | cameras |
|---|---|---|
| `config.json` `input_features` | shape **[6]** | camera1, camera2, **camera3** |
| `policy_preprocessor_step_5_normalizer_processor.safetensors` | mean/std shape **(8,)** | rename step creates only camera1, camera2 |
| training dataset `lerobot/libero` features | shape **[8]** | 2 image keys |

The `[6]` + three cameras are `smolvla_base`'s SO-100 features, carried through verbatim. The checkpoint *works* (SmolVLA pads state to `max_state_dim` and iterates over the images actually present), but anything that trusts `config.json` — integration code, config-driven eval, model cards — reads a wrong spec. We hit this integrating the checkpoint into an external runtime and initially built the wrong observation vector from the config.

### This is the root cause of #2418

The reproduction there is exactly this path (`--policy.path=lerobot/smolvla_base` + `--dataset.repo_id=HuggingFaceVLA/libero` + correct `rename_map`): training runs, **eval crashes with a tensor shape mismatch**, and the workaround that thread converged on — manually passing `--policy.input_features` with the true shapes — is precisely the refresh that L334 skips.

### Proposed fix

Mirror the `output_features` behaviour, with a loud warning on mismatch:

```python
new_inputs = {key: ft for key, ft in features.items() if key not in cfg.output_features}
if cfg.input_features and cfg.input_features != new_inputs:
    logging.warning(
        f"input_features from the pretrained config {cfg.input_features} do not match "
        f"the dataset features {new_inputs}; using the dataset's."
    )
cfg.input_features = new_inputs
```

One design point for maintainers: an **explicit** `--policy.input_features` CLI override (the #2418 workaround) flows into the same `cfg`, so "dataset always wins" would clobber it. If CLI-provided features can be distinguished from checkpoint-loaded ones, precedence should be CLI > dataset > checkpoint; if not, even just warning loudly on the mismatch (instead of silently keeping stale features) plus regenerating the affected official checkpoint configs (`smolvla_libero`: state `[8]`, camera1/camera2 only) would prevent both the #2418 class of failures and the self-contradicting configs.

Happy to send a PR for whichever semantics you prefer.

---
Found while building a dataset/config contract linter (#2326): its "stale stats dims" rule fired on the checkpoint. lerobot 0.4.4, verified unchanged on `main` @ `bf31dd7`.

**Comment by easyrider11** [comment]
Follow-up with ecosystem-scale data: **this is not one checkpoint — it's at least 24 of the top 300.**

I added a checkpoint mode to the dataset linter from #2326 (config.json vs the normalization stats shipped in the same repo; metadata-only — processor stats files are KB-sized, and old-style checkpoints are read via an HTTP Range request on the `model.safetensors` header, no weights downloaded) and swept the 300 most-downloaded `lerobot`-tagged model repos:

- **24 / 251 verifiable checkpoints (9.6%) provably carry this bug**: their own `train_config.json` records a finetune (`policy.pretrained_path` set) while `config.json`'s `observation.state` dims contradict their own shipped normalizer stats.
- **Seven are official `lerobot/` releases** (~38.5k combined downloads), all declaring `smolvla_base`'s `state [6]` + 3 cameras: `smolvla_libero` (8-dim stats), `smolvla_robotwin` (14), `smolvla_libero_plus` (8), `smolvla_metaworld` (4), `smolvla_robocasa` (16), `smolvla_vlabench` (7), `smolvla_robocerebra` (8).
- **Mechanism-specific rate: 39% (19/49) of all `smolvla_base` finetunes in the sample.** The sample even shows second-generation propagation — a stale finetune of a stale finetune.
- 20 further checkpoints declare phantom cameras their preprocessor's rename map never produces.

Deliberately excluded to keep the number conservative: pi0-family repos declaring `state [32]` (`max_state_dim` padding — a distinct, deliberate-looking convention), non-canonical STATE features like pointclouds (per-channel stats conventions make size comparison invalid), and anything without train_config provenance.

Full methodology, per-repo results and reproduce scripts: https://github.com/easyrider11/lerobot-dataset-lint/tree/main/studies/checkpoint-sweep

The `uvx --from git+https://github.com/easyrider11/lerobot-dataset-lint lerobot-lint <repo>` one-liner now auto-detects checkpoints, so affected authors can check their own in seconds. Happy to help fix the seven official configs and/or send the `make_policy` PR discussed above.

**Comment by Marchematics** [comment]
Fixed in #4540.
